pageserver/src/layered_repository.rs

```diff
@@ -2379,19 +2399,28 @@ impl LayeredTimeline {
         let mut layers = self.layers.write().unwrap();
         let mut new_layer_paths = HashSet::with_capacity(new_layers.len());
         for l in new_layers {
-            new_layer_paths.insert(l.path());
+            let new_delta_path = l.path();
+
+            // update the timeline's physical size
+            self.physical_size
+                .fetch_add(new_delta_path.metadata()?.len(), atomic::Ordering::SeqCst);
+
+            new_layer_paths.insert(new_delta_path);
             layers.insert_historic(Arc::new(l));
         }
 
         // Now that we have reshuffled the data to set of new delta layers, we can
         // delete the old ones
         let mut layer_paths_do_delete = HashSet::with_capacity(deltas_to_compact.len());
-        for l in &deltas_to_compact {
-            l.delete()?;
+        drop(all_keys_iter);
+        for l in deltas_to_compact {
```

Comment:
Previously, `deltas_to_compact` needs to be borrowed here because `all_keys_iter` still exists. I'm not sure why Rust doesn't detect that `all_keys_iter` is not referred till the end the function to allow to move `deltas_to_compact`, so I add an explicit drop here.

@knizhnik can you review this change?

Reply:
> I'm not sure why Rust doesn't detect that all_keys_iter is not referred till the end the function to allow to move deltas_to_compact,
Most likely it is because `all_keys_iter` iterator has destructor which according to the specification should be invoced at the end of the scope. And running the destructor is considered a use ‒ obviously the last one.
For example the following fragment of code is also not compiled:

```
let mut data = vec![1, 2, 3];
let x = &data[0];
println!("{}", x);
// This is OK, x is no longer needed
data.push(4);

```

So your change is ok, 
But I just wonder why new version is better than old one?

Reply:
> But I just wonder why new version is better than old one?

`deltas_to_compact` is moved, so we don't need to clone in `layers.remove_historic(l.clone());`. I know `Arc` clone is not that expensive, but it's still nice not to have that.